### PR TITLE
Fixes to ensure streams are correctly mapped to plots

### PR DESCRIPTION
**Pull request by philippjfr**
This fix addresses a tricky situation where nested overlays were not resolved correctly because it would simply keep recursing until it got to the last layer. This meant the resulting subplot(s) would not update. We now still do the recursing (though if I spent a lot more time on this we maybe wouldn't have to do so), but also record all the streams along the way.

Also fixes a separate bug where during zorder assignment we did not take into account DynamicMaps that had an (Nd)Overlay as input, which would confuse the zorder assignment and therefore meant that streams would be assigned to the wrong plot (or no plot at all). 

Fixes https://github.com/holoviz/holoviews/issues/5595
Fixes https://github.com/holoviz/holoviews/issues/5434

- [x] Add test

### CODE REVIEW

**Review by maximlt — APPROVED** [review]
Very happy to see https://github.com/holoviz/holoviews/issues/5595 being fixed! Overall LGTM (made some minor optional suggestions).

I was not able to see whether the OP example in https://github.com/holoviz/holoviews/issues/5434 is fixed as there seems to be some issues with the box edit tool, I can't create any selection.

@philippjfr make sure the PR title is updated based on your last changes since it seems the recursion is still done.

Technically `spit_dmap_overlay` is public as documented on the website (https://holoviews.org/reference_manual/holoviews.plotting.html#holoviews.plotting.util.split_dmap_overlay) so that's a breaking change. No one's probably using so I don't think it's an issue, the release notes might just include a comment on that.

**Comment by philippjfr** [comment]
>I was not able to see whether the OP example in #5434 is fixed as there seems to be some issues with the box edit tool, I can't create any selection.

The UX of drawing boxes has changed in Bokeh (in quite a non-obvious way). You have to hold down shift and then draw.

File: holoviews/tests/plotting/test_plotutils.py [hunk 1]
```diff
@@ -364,61 +371,61 @@ def test_dmap_ndoverlay(self):
         test = self.dmap_ndoverlay
         initialize_dynamic(test)
         layers = [self.dmap_ndoverlay, self.dmap_ndoverlay]
-        self.assertEqual(split_dmap_overlay(test), layers)
+        self.assertEqual(split_dmap_overlay(test)[0], layers)
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply holoviews/tests/plotting/test_plotutils.py:374]: None of these actually have any streams, will have to add a new test.

File: holoviews/tests/plotting/test_plotutils.py [hunk 2]
```diff
@@ -202,6 +202,13 @@ def test_dynamic_compute_overlayable_zorders_mixed_dynamic_and_non_dynamic_ndove
         self.assertIn(curve, sources[2])
         self.assertNotIn(ndoverlay, sources[2])
 
+    def test_dynamic_compute_overlayable_zorders_ndoverlays_as_input(self):
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply holoviews/tests/plotting/test_plotutils.py:205]: Had real trouble finding a smaller reproducible example for the first issue.

**Comment by philippjfr** [comment]
I'm going to merge this and follow up with tests in a bit as I have to get the RC out asap.

**Pull request merged**